PHP Process Manager

adamziel opened the pull request:
Adds a PHP Process manager, a class that can spawn maintain a number of PHP instances at the same time.

The idea is to have:

* A single "primary" PHP instance that's never killed – it contains the reference filesystem used by all other PHP instances.
* A pool of disposable PHP instances that are spawned to handle a single request and reaped immediately after.

Spawning new PHP instances is reasonably fast and can happen on demand, there's no need to keep a pool of "warm" instances that occupies the memory all the time.

Example usage:

```ts
const mgr = new PHPProcessManager({
	phpFactory: async () => NodePHP.load(RecommendedPHPVersion),
	maxPhpInstances: 4,
});

const instance = await mgr.getInstance();
await instance.php.run({ code });
instance.reap();

```

Related to https://github.com/WordPress/wordpress-playground/pull/1287

 ## Testing instructions

Confirm the unit tests pass. This PR only adds new code, it does not plug in the PHPProcessManager class into any request dispatching code yet.

## What is this PR doing?

## What problem is it solving?

## How is the problem addressed?

## Testing Instructions

adamziel commented on packages/php-wasm/universal/src/lib/php-process-manager.ts:
```diff
@@ -0,0 +1,182 @@
+		} catch (error) {
+			throw new MaxPhpInstancesError(this.maxPhpInstances);
+		}
+		const php = await this.phpFactory!(factoryArgs);
+		return {
+			php,
+			reap() {
+				php.exit();
+				release();
+			},
+		};
```
```suggestion
		try {
			const php = await this.phpFactory!(factoryArgs);
			return {
				php,
				reap() {
					php.exit();
					release();
				},
			};
		} catch(e) {
			release();
			throw e;
		}
```

adamziel commented:
cc @dmsnell @brandonpayton @bgrgicak for reviews – this is an important component and I'd love to discuss it before moving further.

bgrgicak commented:
> this is an important component and I'd love to discuss it before moving further.

The approach makes sense to me. Do you have any specific concerns?

adamziel commented:
> The approach makes sense to me. Do you have any specific concerns?

Nothing specific, I just wanted to make sure I get some really smart folks to take a look :-)

brandonpayton commented on packages/php-wasm/universal/src/lib/php-process-manager.ts:
```diff
@@ -0,0 +1,233 @@
+
+	/**
+	 * Initiated spawning of a new PHP instance.
+	 * This function is not async on purpose – it needs to synchronously
```
Instead of "not async", maybe we can just say "synchronous". Initially, "not async on purpose" made me wonder if the sentence was implying "unintentionally async", and I think it is clearer to say "sync" than "not not-sync".

adamziel replied:
good call!

brandonpayton commented on packages/php-wasm/universal/src/lib/php-process-manager.ts:
```diff
@@ -0,0 +1,233 @@
+		}
+	}
+
+	async [Symbol.dispose]() {}
```
Based on the proposal's example here:
https://github.com/tc39/proposal-explicit-resource-management/blob/6277c43005e2f2b366a21594592a8d5b4dd4759e/README.md#L391-L392

I don't think it is necessary to have both 
`[Symbol.dispose]() {}`
and 
`async [Symbol.asyncDispose]() {`

Since the one with sync semantics is empty, is it fine to just remove it?

adamziel replied:
I had to switch to extending AsyncDisposable instead of Disposable, but that did the trick

brandonpayton commented on packages/php-wasm/universal/src/lib/php-process-manager.ts:
```diff
@@ -0,0 +1,233 @@
+		try {
+			release = await this.semaphore.acquire();
+		} catch (error) {
+			throw new MaxPhpInstancesError(this.maxPhpInstances);
```
It might feel too verbose, but perhaps it would be good to check explicitly for the expected exception type and rethrow for any others. Otherwise, an error other than `AcquireTimeoutError` will be hidden and masquerade as a max-instances condition.

adamziel replied:
Good call!

brandonpayton commented on packages/php-wasm/universal/src/lib/php-process-manager.ts:
```diff
@@ -0,0 +1,233 @@
+	 *
+	 * @throws {Error} when called twice before the first call is resolved.
+	 */
+	async getPrimaryPhp() {
```
What is the purpose of being able to get the primary PHP instance rather than just calling `getInstance()` to get *some* PHP instance?

adamziel replied:
Oftentimes we just want to call `mkdir` or `readFile` and it would be a waste to spin a new instance just to do that. Plus you can only do certain FS operations, like new mounts, on the primary instance.

brandonpayton commented on packages/php-wasm/universal/src/lib/php-process-manager.ts:
```diff
@@ -0,0 +1,233 @@
+	 * @throws {MaxPhpInstancesError} when the maximum number of PHP instances is reached
+	 *                                and the waiting timeout is exceeded.
+	 */
+	async getInstance(): Promise<SpawnedPHP<PHP>> {
```
The name `getInstance()` reminds me of Singleton methods like `Something::instance()` or `Something::getInstance()`, and it might be clearer to name the method more specifically like `getPhpInstance()`, `requestPhpInstance()`, `acquirePhpInstance()`.

adamziel replied:
`acquirePHPInstance` it is

brandonpayton commented on packages/php-wasm/universal/src/lib/php-process-manager.ts:
```diff
@@ -0,0 +1,233 @@
+		super(
+			`Requested more concurrent PHP instances than the limit (${limit}).`
+		);
+		this.name = 'MaxPhpInstancesError';
```
Random aside about JS:
It would be convenient if Error would automatically set `this.name` to `this.constructor.name`. That would probably be too magical/clever, but it seems strange to have to manually set it for all Error subtypes.

adamziel replied:
yes! I changed the hardcoded string to this.constructor.name, though, so thank you for that

brandonpayton reviewed (commented):
@adamziel, here are my thoughts and questions so far. I may keep reading and thinking about this for a bit but want to get some feedback out there.

In addition to the inline feedback, some other questions are:
- Does there need to be a primary PHP instance?
- What happens if a consumer disposes of the primary PHP instance? Does it matter?
- Would it make sense to track parent/child relationships? Maybe that is something best left to the consumer of the PHP Process Manager.

adamziel commented:
All good notes @brandonpayton, thank you!

> Does there need to be a primary PHP instance?

https://github.com/WordPress/wordpress-playground/pull/1301#discussion_r1578420516

> What happens if a consumer disposes of the primary PHP instance? Does it matter?

They don't get access to a `reap()` function that would actually call `php.exit()`. If they do call `php.exit()` manually, their program will break. It's a footgun for sure, but I'm okay with responding to that "well just don't call exit() then" :-) If it bites us at one point, we can add a check like `if(primary.hasExited()) { throw new Error("The primary PHP instance have exited and this PHPProcessManager can no longer be used. Make sure to avoid calling php.exit(), that's managed for you. "); }`

> Would it make sense to track parent/child relationships? Maybe that is something best left to the consumer of the PHP Process Manager.

I was thinking about that and here's what I realized:

* It would be handy to kill the entire child process group, as in Erlang's supervisor trees.
* We can track that when PHP spawns another one directly, as in `proc_open()`, but we cannot easily track that when PHP issues a loopback request via HTTP and that request spawns another process – in this case it would technically be a child of the request handler.
* At the same time, the request handler isn't a WASM module in the same sense as PHP so now we'd have to introduce more OS-like primitives with a generic process interface, a process table or a supervisor tree, potentially implement file descriptors outside of WASM context etc. I'd rather avoid that complexity while we can. Eventually perhaps we could just build a real process manager to WASM, add custom handlers, and get a lot of that for free?

adamziel commented:
The unit test failure is an intermittent timeout. I've addressed the feedback and I'll go ahead and merge this one. Thank you so much for the discussion!
